Upgrade flexmailer symfony code to run on v3.4 and 4.4 as per core PR…

seamuslee001 opened the pull request:
… #13780

ping @MikeyMJCO @totten @karing

seamuslee001 committed:
Upgrade flexmailer symfony code to run on v3.4 and 4.4 as per core PR #13780

totten committed:
(REF) Remove dead code (`registerListeners()` et al)

`CIVICRM_FLEXMAILER_HACK_LISTENERS` and `registerListeners()` function were
used in older versions to wire-up Flexmailer's dispatchers.  This is now
accomplished via public API, `hook_civicrm_container` - seems to work, no
reason to go back.

Note: I was confused by two things:

1. The previous calls didn't look like they should work.

2. Somehow, the API override still worked - even though there was no obvious callpath to `addSubscriberService()`
   It turns out that `registerApiProvider()` handles registration with the dispatcher, so `addSubscriberService()`
   isn't actually used.

totten commented on src/Services.php:
```diff
@@ -50,34 +52,34 @@ public static function registerServices(ContainerBuilder $container) {
-      $dispatcher->addListenerService($listenerSpec[0], $listenerSpec[1], $listenerSpec[2]);
-      $dispatcher->addSubscriberService('civi_flexmailer_api_overrides', 'Civi\FlexMailer\API\Overrides');
+      $dispatcher->addListener($listenerSpec[0], [new SercviceClosureArgument(new Reference($listenerSpec[1][0])), $listenerSpec[1][1]], $listenerSpec[2]);
+      $dispatcher->addSubscriber(new Reference('civi_flexmailer_api_overrides'));
```
Aah, so this example change is much clearer than the upstream docs. :+1: I'm not certain that these lines work per se (more in a sec), but it was enlightening to read.

Digging in on this PR/function was a bit of rabbit hole...

* Neither the [upstream docs](https://symfony.com/doc/4.4/components/event_dispatcher/container_aware_dispatcher.html), [changelog](https://github.com/symfony/symfony/blob/3.3/CHANGELOG-3.3.md), [deprecation PR](https://github.com/symfony/symfony/pull/20937), nor [docs issue](https://github.com/symfony/symfony/issues/22402) provide an example snippet or reference to the alternative helpers. Kudos to you for figuring a work-a-like in Symfony 4.x's class model.
* There's a typo in the class name (`Sercvice`). Digging into this revealed that the function `registerListeners()` is never actually called. It's leftover from some alpha/beta hackery and was disabled (17f142e1539ac2e507d35b837607b8fad965b6a5) when `hook_civicrm_container` became more reliable. This also explains things worked despite some more flaws in the old variant of `registerListeners()`.
* How does `civi_flexmailer_api_overrides` work *at all*? Well, it does. ✅  It's because there's a call to `registerApiProvider()`, and *that* is what really handles registration.

totten committed:
Re-enable usage of `addListenerService()`

In core's #17380, we ultimately preserved `addListenerService()`, so this doesn't have to be removed.

totten commented:
Since the core PR restored `addListenerService()`, I've updated/simplified those bits. The calls to `setPublic()` seem to be backward/forward compatible. Re-ran the test suite, and it passes on core@5.13. :)
